src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,190 @@
+// <copyright file="TagTransformerJsonHelper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Buffers;
+using System.Text;
+using System.Text.Json;
+
+namespace OpenTelemetry.Internal;
+
+#nullable enable
+
+internal static class TagTransformerJsonHelper
+{
+    internal static string JsonSerializeArrayTag(Array array)
+        => array switch
+        {
+            // This switch needs to support the same set of types as those in TagTransformer.TransformArrayTagInternal
+            char[] charArray => WriteToString(charArray, WriteCharValue),
+            string[] stringArray => WriteToString(stringArray, WriteStringValue),
+            bool[] booleanArray => WriteToString(booleanArray, WriteBooleanValue),
+
+            // Runtime allows casting byte[] to sbyte[] and vice versa, so pattern match to sbyte[] doesn't work since it goes through byte[]
+            // Similarly (array is sbyte[]) doesn't help either. Only real type comparison works.
+            // This is true for byte/sbyte, short/ushort, int/uint and so on
+            byte[] byteArray => array.GetType() == typeof(sbyte[]) ? WriteToString((sbyte[])array, WriteSByteValue) : WriteToString(byteArray, WriteByteValue),
```

Comment:
`byte[]` serialize to base64 strings with JsonSerializer.

![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/8291187/3368b994-d499-47b4-9cba-977da4af49c2)

Reply:
I thought so, but look at the tests - it uses the reflection based serializer as the "expected" value. Originally I actually coded it to produce a Base64 string, but the tests failed, because the reflection based serializer didn't do it.

I'll play with it a bit more - probably will try to debug into the existing OTel code to see what it will end up doing in real.

Reply:
![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/10670590/e4fc84f1-214b-46f2-b98d-d03f49c1c057)

I think OTel ended up always calling the method with `Array` typed parameter, so it got the second behavior (which is to write it as an array of numbers).

Reply:
This is a side-effect of hardcoding Base64 encoding specifically for values of declared type `byte[]`. It results in different outputs if the same value is upcast to a different type:
```C#
byte[] data = new byte[] { 1 };

Console.WriteLine(JsonSerializer.Serialize(data)); // "AQ=="
Console.WriteLine(JsonSerializer.Serialize<IEnumerable<byte>>(data)); // [1]
Console.WriteLine(JsonSerializer.Serialize<Array>(data)); // [1]
```